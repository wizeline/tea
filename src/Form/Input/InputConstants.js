// @flow

const inputTypes = {
  TEXT: 'text',
  PASSWORD: 'password',
};

export default inputTypes;
