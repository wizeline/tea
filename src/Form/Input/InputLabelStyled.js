// @flow

import styled from 'styled-components';
import defaultErrorTheme from '../../themes/defaultTheme/components/Error';
import defaultTextTheme from '../../themes/defaultTheme/components/Text';
import {
  otherConstants,
  textConstants,
} from '../../themes/defaultTheme/constants';
import { getThemeProp } from '../../utils';

const BaseLabelStyled = styled.label`
  display: block;
  font-size: 0.75rem;
`;

const LabelStyled = BaseLabelStyled.extend`
  color: ${getThemeProp(textConstants.TEXT_INPUT)};
  margin: 0 0 0.5rem;
`;

const ErrorLabelStyled = BaseLabelStyled.extend`
  color: ${getThemeProp(otherConstants.TEXT_ERROR)};
  visibility: ${({ error }) => (error ? 'visible' : 'hidden')};
  margin: 0.125rem 0 0 0;
`;

const InputContainer = styled.div`
  position: relative;
  & > div {
    position: absolute;
    left: 1rem;
    /*
      To centre the Icon we have to use calc, as the current layout
      is something like this:
        <InputContainer>
          <TopLabel />
          <Input />
          <Icon />
          <BottomLabel />
        </InputContainer>
      The height of the InputContainer is 78px because:
        TopLabel height + margin = 14px + 8px = 22px
        Input height = 40px
        BottomLabel height + margin = 14px + 2px = 16px
        Thus, 22px + 40px +  16px = 78px
      To centre the Icon we push it down by 50%
        then we have to pull it up by half of the TopLabel's height, 11px
        after that, we push it down by half of the BottomLabel's height, 8px
        so, at the end, we have 50% + 11px - 8px = 50% + 3px
        which in rem is 50% + 0.1875rem
    */
    top: calc(50% + 0.1875rem);
    transform: translateY(-50%);

    & > svg {
      fill: ${getThemeProp(textConstants.TEXT_INPUT_PLACEHOLDER)};
    }
  }
`;

LabelStyled.defaultProps = {
  theme: {
    ...defaultTextTheme,
  },
};

ErrorLabelStyled.defaultProps = {
  theme: {
    ...defaultErrorTheme,
  },
};

export { LabelStyled, ErrorLabelStyled, InputContainer };
