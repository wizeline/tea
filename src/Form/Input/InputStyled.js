// @flow

import { placeholder } from 'polished';
import styled from 'styled-components';
import defaultInputTheme from '../../themes/defaultTheme/components/Input';
import {
  borderConstants,
  textConstants,
} from '../../themes/defaultTheme/constants';
import { getThemeProp } from '../../utils';
import type { InputProps } from './InputTypes';

const getWidth = ({ size }: InputProps): number => size || 18.75;
const getPlaceholder = (prop: string) => (props: InputProps) =>
  placeholder({ color: getThemeProp(prop)(props) });

const BaseInputStyled = styled.input`
  border-radius: 0.1875rem;
  box-sizing: border-box;
  color: ${getThemeProp(textConstants.TEXT_BODY)};
  font-size: 0.875rem;
  height: 2.5rem;
  outline: none;
  padding: 0 1rem;
  transition: all 0.1s;
  width: ${getWidth}rem;
  ${getPlaceholder(textConstants.TEXT_INPUT_PLACEHOLDER)};
`;

const getInputBorderVariant = (
  border,
  borderFocus = border,
) => BaseInputStyled.extend`
  border: solid 0.0625rem ${getThemeProp(border)};

  &:focus {
    border: solid 0.0625rem ${getThemeProp(borderFocus)};
  }
`;

const InputStyled = getInputBorderVariant(
  borderConstants.BORDER_INPUT,
  borderConstants.BORDER_INPUT_FOCUS,
);

const ErrorInputStyled = getInputBorderVariant(
  borderConstants.BORDER_INPUT_INVALID,
);

BaseInputStyled.defaultProps = {
  theme: {
    ...defaultInputTheme,
  },
};

export { InputStyled, ErrorInputStyled };
