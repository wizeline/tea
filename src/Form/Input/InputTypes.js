// @flow

import type { Node } from 'react';

export type BaseInputProps = {
  autofocus: boolean,
  disabled: boolean,
  name: string,
  onChange: () => void,
  readOnly: boolean,
  required: boolean,
  type: string,
  value: string,
  theme: {
    borderInput: string,
    borderInputFocus: string,
    borderInputInvalid: string,
    borderSecondary: string,
    borderSecondaryDisabled: string,
    borderSecondaryHover: string,
    textBody: string,
    textDanger: string,
    textDangerDisabled: string,
    textDangerHover: string,
    textHeading: string,
    textIcon: string,
    textIconInvert: string,
    textInput: string,
    textInputFocus: string,
    textInputInvalid: string,
    textInputPlaceholder: string,
    textPrimary: string,
    textPrimaryDisabled: string,
    textPrimaryHover: string,
    textPrimaryInvert: string,
    textSecondary: string,
    textSecondaryDisabled: string,
    textSecondaryHover: string,
    textSubtitle: string,
  },
};

export type InputProps = BaseInputProps & {
  autocomplete: string,
  error: boolean,
  errorMessage: string,
  label: string,
  maxLenght: number,
  placeholder: string,
  size: number,
  icon: string,
};

export type InputLabelProps = {
  children: Node,
  error?: boolean,
  errorMessage?: string,
  label?: string,
};
