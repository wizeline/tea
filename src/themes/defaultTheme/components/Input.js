// @flow

import defaultTextTheme from '../../defaultTheme/components/Text';
import borders from '../../styles/borders';

export default {
  ...defaultTextTheme,
  ...borders,
};
