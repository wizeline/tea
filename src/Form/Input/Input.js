// @flow

import React from 'react';
import Icon from '../../icons/Icon';
import inputTypes from './InputConstants';
import {
  ErrorLabelStyled,
  InputContainer,
  LabelStyled,
} from './InputLabelStyled';
import { ErrorInputStyled, InputStyled } from './InputStyled';
import type { InputProps, InputLabelProps } from './InputTypes';

const InputWithLabels = ({
  children,
  error,
  label,
  errorMessage,
}: InputLabelProps) => (
  <InputContainer>
    <LabelStyled>{label}</LabelStyled>
    {children}
    <ErrorLabelStyled error={error}>{errorMessage}</ErrorLabelStyled>
  </InputContainer>
);

const Input = (props: InputProps) => {
  const { error, label, errorMessage, onChange, type, readOnly, icon } = props;
  const typeStr =
    type && type === inputTypes.PASSWORD
      ? inputTypes.PASSWORD
      : inputTypes.TEXT;
  const FinalInput = error ? ErrorInputStyled : InputStyled;
  return (
    <InputWithLabels error={error} label={label} errorMessage={errorMessage}>
      <FinalInput
        {...props}
        onchange={onChange}
        type={typeStr}
        readonly={readOnly}
      />
      <Icon xsmall name={icon} />
    </InputWithLabels>
  );
};

export default Input;
