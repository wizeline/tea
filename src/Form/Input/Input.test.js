import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Input from './Input';

describe('Input', () => {
  it('functions correctly without props', () => {
    const wrapper = shallow(<Input />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('functions correctly with props', () => {
    const props = {
      placeholder: 'Placeholder',
      label: 'Text input label',
      errorMessage: 'Text input error message',
      error: true,
      type: 'password',
    };
    const wrapper = shallow(<Input {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('has the correct size', () => {
    const props = {
      placeholder: 'Placeholder',
      label: 'Text input label',
      errorMessage: 'Text input error message',
      size: 15,
    };
    const wrapper = mount(<Input {...props} />);
    expect(wrapper.find('input').prop('size')).toBe(15);
  });

  it('shows the error message when necessary', () => {
    const props = {
      placeholder: 'Placeholder',
      label: 'Text input label',
      errorMessage: 'Text input error message',
      error: true,
    };
    const wrapper = mount(<Input {...props} />);
    expect(wrapper.find('label').last()).toHaveStyleRule(
      'visibility',
      'visible',
    );
  });

  it('functions correctly with Icon', () => {
    const props = {
      placeholder: 'Placeholder',
      label: 'Text input label',
      errorMessage: 'Text input error message',
      error: true,
      icon: 'dots-h',
    };
    const wrapper = mount(<Input {...props} />);
    expect(wrapper.find('input')).toHaveStyleRule('padding', '0 1rem 0 2.5rem');
  });

  it('calls onChange function', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Input onChange={onChange} />);
    wrapper.find('input').simulate('change');
    expect(onChange.mock.calls.length).toEqual(1);
  });
});
