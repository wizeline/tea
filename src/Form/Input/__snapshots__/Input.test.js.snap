// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input functions correctly with props 1`] = `
<InputWithLabels
  error={true}
  errorMessage="Text input error message"
  label="Text input label"
>
  <styled.input
    error={true}
    errorMessage="Text input error message"
    label="Text input label"
    placeholder="Placeholder"
    theme={
      Object {
        "borderInput": "#E6E7E8",
        "borderInputFocus": "#50A5E7",
        "borderInputInvalid": "#ED524C",
        "borderSecondary": "#8B9193",
        "borderSecondaryDisabled": "#E6E7E8",
        "borderSecondaryHover": "#50A5E7",
        "textBody": "#3F484B",
        "textDanger": "#FFFFFF",
        "textDangerDisabled": "#CDD0D2",
        "textDangerHover": "#FFFFFF",
        "textError": "#ED524C",
        "textHeading": "#3F484B",
        "textIcon": "#FFFFFF",
        "textIconInvert": "#222729",
        "textInput": "#3F484B",
        "textInputFocus": "#8B9193",
        "textInputInvalid": "#E92720",
        "textInputPlaceholder": "#B2B5B7",
        "textPrimary": "#3F484B",
        "textPrimaryDisabled": "#8B9193",
        "textPrimaryHover": "#FFFFFF",
        "textPrimaryInvert": "#FFFFFF",
        "textSecondary": "#3F484B",
        "textSecondaryDisabled": "#CDD0D2",
        "textSecondaryHover": "#FFFFFF",
        "textSubtitle": "#8B9193",
      }
    }
    type="password"
  />
  <Icon
    theme={
      Object {
        "backgroundDanger": "#ED524C",
        "backgroundDangerDisabled": "#E6E7E8",
        "backgroundInput": "#FFFFFF",
        "backgroundPrimary": "#50A5E7",
        "backgroundPrimaryDisabled": "#E6E7E8",
        "backgroundSecondary": "#FFFFFF",
        "backgroundSecondaryDisabled": "#FFFFFF",
        "backgroundSecondaryHover": "#FFFFFF",
        "textBody": "#3F484B",
        "textDanger": "#FFFFFF",
        "textDangerDisabled": "#CDD0D2",
        "textDangerHover": "#FFFFFF",
        "textHeading": "#3F484B",
        "textIcon": "#FFFFFF",
        "textIconInvert": "#222729",
        "textInput": "#3F484B",
        "textInputFocus": "#8B9193",
        "textInputInvalid": "#E92720",
        "textInputPlaceholder": "#B2B5B7",
        "textPrimary": "#3F484B",
        "textPrimaryDisabled": "#8B9193",
        "textPrimaryHover": "#FFFFFF",
        "textPrimaryInvert": "#FFFFFF",
        "textSecondary": "#3F484B",
        "textSecondaryDisabled": "#CDD0D2",
        "textSecondaryHover": "#FFFFFF",
        "textSubtitle": "#8B9193",
      }
    }
    xsmall={true}
  />
</InputWithLabels>
`;

exports[`Input functions correctly without props 1`] = `
<InputWithLabels>
  <styled.input
    theme={
      Object {
        "borderInput": "#E6E7E8",
        "borderInputFocus": "#50A5E7",
        "borderInputInvalid": "#ED524C",
        "borderSecondary": "#8B9193",
        "borderSecondaryDisabled": "#E6E7E8",
        "borderSecondaryHover": "#50A5E7",
        "textBody": "#3F484B",
        "textDanger": "#FFFFFF",
        "textDangerDisabled": "#CDD0D2",
        "textDangerHover": "#FFFFFF",
        "textError": "#ED524C",
        "textHeading": "#3F484B",
        "textIcon": "#FFFFFF",
        "textIconInvert": "#222729",
        "textInput": "#3F484B",
        "textInputFocus": "#8B9193",
        "textInputInvalid": "#E92720",
        "textInputPlaceholder": "#B2B5B7",
        "textPrimary": "#3F484B",
        "textPrimaryDisabled": "#8B9193",
        "textPrimaryHover": "#FFFFFF",
        "textPrimaryInvert": "#FFFFFF",
        "textSecondary": "#3F484B",
        "textSecondaryDisabled": "#CDD0D2",
        "textSecondaryHover": "#FFFFFF",
        "textSubtitle": "#8B9193",
      }
    }
    type="text"
  />
  <Icon
    theme={
      Object {
        "backgroundDanger": "#ED524C",
        "backgroundDangerDisabled": "#E6E7E8",
        "backgroundInput": "#FFFFFF",
        "backgroundPrimary": "#50A5E7",
        "backgroundPrimaryDisabled": "#E6E7E8",
        "backgroundSecondary": "#FFFFFF",
        "backgroundSecondaryDisabled": "#FFFFFF",
        "backgroundSecondaryHover": "#FFFFFF",
        "textBody": "#3F484B",
        "textDanger": "#FFFFFF",
        "textDangerDisabled": "#CDD0D2",
        "textDangerHover": "#FFFFFF",
        "textHeading": "#3F484B",
        "textIcon": "#FFFFFF",
        "textIconInvert": "#222729",
        "textInput": "#3F484B",
        "textInputFocus": "#8B9193",
        "textInputInvalid": "#E92720",
        "textInputPlaceholder": "#B2B5B7",
        "textPrimary": "#3F484B",
        "textPrimaryDisabled": "#8B9193",
        "textPrimaryHover": "#FFFFFF",
        "textPrimaryInvert": "#FFFFFF",
        "textSecondary": "#3F484B",
        "textSecondaryDisabled": "#CDD0D2",
        "textSecondaryHover": "#FFFFFF",
        "textSubtitle": "#8B9193",
      }
    }
    xsmall={true}
  />
</InputWithLabels>
`;
